test-harness/src/internal.rs

```diff
@@ -0,0 +1,225 @@
+            );
+            cleanup().await;
+
+            if err.is_panic() {
+                std::panic::resume_unwind(err.into_panic());
+            } else {
+                panic!("Test cancelled? {err:?}");
+            }
+        }
+        Ok(()) => {
+            cleanup().await;
+        }
+    };
+}
+
+/// Save the test logs in `TEMPDIR` to `cargo_target_tmpdir`.
+///
+/// This is called when a `#[test]`-annotated function panics, to persist the logs for further
+/// analysis.
+fn save_test_logs(test_name: String, cargo_target_tmpdir: PathBuf) {
+    let log_path: PathBuf = TEMPDIR.with(|tempdir| {
+        tempdir
+            .borrow()
+            .as_deref()
+            .map(|path| path.join(crate::ghcid_ng::LOG_FILENAME))
+            .expect("`test_harness::TEMPDIR` is not set")
+    });
+
+    let test_name = test_name.replace("::", "-");
+    let persist_log_path = cargo_target_tmpdir.join(format!("{test_name}.json"));
+    if persist_log_path.exists() {
+        // Cargo doesn't manage `CARGO_TARGET_TMPDIR` for us, so we remove output from old tests
+        // ourself.
+        std::fs::remove_file(&persist_log_path).expect("Failed to remove log output");
+    }
+
+    if !log_path.exists() {
+        eprintln!("No logs were written");
+    } else {
+        let logs = std::fs::read_to_string(log_path).expect("Failed to read logs");
+        std::fs::write(&persist_log_path, logs).expect("Failed to write logs");
+        eprintln!("Wrote logs to {}", persist_log_path.display());
+    }
+}
+
+/// Perform end-of-test cleanup.
+///
+/// 1. Kill the [`GHC_PROCESS`].
+/// 2. Remove the [`TEMPDIR`] from the filesystem.
+async fn cleanup() {
+    let child = GHC_PROCESS.with(|child| child.take());
+    match child {
+        None => {
+            panic!("`GHC_PROCESS` is not set");
+        }
+        Some(mut child) => {
+            send_signal(&child, Signal::SIGINT).expect("Failed to send SIGINT to `ghcid-ng`");
+            match tokio::time::timeout(Duration::from_secs(10), child.wait()).await {
+                Err(_) => {
+                    tracing::info!("ghcid-ng didn't exit in time, killing");
+                    child
+                        .kill()
+                        .await
+                        .expect("Failed to kill `ghcid-ng` after test completion");
+                }
+                Ok(Ok(status)) => {
+                    tracing::info!(?status, "ghcid-ng exited");
+                }
+                Ok(Err(err)) => {
+                    tracing::error!("Waiting for ghcid-ng to exit failed: {err}");
+                }
+            }
+        }
+    }
+
+    let path = TEMPDIR.with(|path| path.take());
+    match path {
+        None => {
+            panic!("`TEMPDIR` is not set");
+        }
+        Some(path) => {
+            if let Err(err) = tokio::fs::remove_dir_all(&path).await {
+                // Run `find` on the directory so we can see what's in it?
```

Comment:
Have you had to make use of this fallback code path for removing the directory?  In other words, has `rm -rf` succeeded where `remove_dir_all` failed?